File: types/whale.ts
/**
 * Mirrors the backend's Whale Engine response shape (see
 * backend/app/schemas/whale.py). Transfers are Etherscan-tracked (ETH +
 * a small set of ERC-20 tokens) touching a curated list of known exchange
 * wallets, classified deterministically — see
 * backend/app/intelligence/whales/classifier.py.
 */
export type WhaleWalletType = "EXCHANGE" | "UNKNOWN";
export type WhaleDirection = "TO_EXCHANGE" | "FROM_EXCHANGE";

export interface WhaleTransaction {
  id: string;
  asset: string;
  amount: number;
  usdValue: number;
  walletType: WhaleWalletType;
  direction: WhaleDirection;
  exchange: string | null;
  confidence: number;
  fromAddress: string;
  toAddress: string;
  txHash: string;
  timestamp: string;
}

File: backend/app/schemas/whale.py
from typing import Literal

from app.schemas.base import CamelModel

WalletType = Literal["EXCHANGE", "UNKNOWN"]
WhaleDirection = Literal["TO_EXCHANGE", "FROM_EXCHANGE"]


class WhaleTransaction(CamelModel):
    id: str
    asset: str
    amount: float
    usd_value: float
    wallet_type: WalletType
    direction: WhaleDirection
    exchange: str | None
    confidence: float
    from_address: str
    to_address: str
    tx_hash: str
    timestamp: str
